File: sdk/typescript/src/compliance.ts
/**
 * Compliance Monitor — per UAOP article.
 *
 * Each check returns `true` when the article is *satisfied* for a given call.
 * Checks are intentionally shallow — the SDK ships with a reference rule set;
 * operators extend as needed.
 */
import type { ComplianceStatus } from "./types";

export type UaopArticle = 1 | 2 | 3 | 4 | 5 | 6 | 7;

export interface CallContext {
  principal?: { role?: string; id?: string };
  declaredCapabilities?: string[];
  requestedAction?: string;
  isIrreversible?: boolean;
  hasAuthToken?: boolean;
  humanAskedIdentity?: boolean;
  agentDisclosed?: boolean;
  usedTokens?: number;
  budgetTokens?: number;
  pii?: boolean;
  piiAuthorized?: boolean;
  claimConfidence?: number;
  domain?: string;
}

export interface ComplianceCheck {
  article: UaopArticle;
  passed: boolean;
  rule: string;
  note?: string;
}

export function check(ctx: CallContext): ComplianceCheck[] {
  const out: ComplianceCheck[] = [];

  // Article 1 — Honest Representation
  out.push({
    article: 1,
    rule: "Claims must carry a confidence score >= 0.3 (refuse-below-threshold elsewhere).",
    passed: ctx.claimConfidence === undefined || ctx.claimConfidence >= 0.3,
  });

  // Article 2 — Scope Adherence
  const inScope =
    !ctx.requestedAction ||
    !ctx.declaredCapabilities ||
    ctx.declaredCapabilities.some((c) => ctx.requestedAction!.includes(c));
  out.push({ article: 2, passed: inScope, rule: "Requested action must match a declared capability." });

  // Article 3 — Conflict Escalation (best-effort: no principal conflict field; pass)
  out.push({ article: 3, passed: true, rule: "No known conflict among principals (SDK-level check)." });

  // Article 4 — Irreversibility Caution
  const irreversibleOk = !ctx.isIrreversible || ctx.hasAuthToken === true;
  out.push({
    article: 4,
    passed: irreversibleOk,
    rule: "Irreversible actions require an explicit authorization token.",
  });

  // Article 5 — Resource Etiquette
  const withinBudget = !ctx.budgetTokens || (ctx.usedTokens ?? 0) <= ctx.budgetTokens;
  out.push({ article: 5, passed: withinBudget, rule: "Stay within per-task token budget." });

  // Article 6 — Disclosure
  const disclosed = !ctx.humanAskedIdentity || ctx.agentDisclosed === true;
  out.push({ article: 6, passed: disclosed, rule: "Disclose AI nature when directly asked." });

  // Article 7 — Data Minimization
  const piiOk = !ctx.pii || ctx.piiAuthorized === true;
  out.push({ article: 7, passed: piiOk, rule: "Do not transmit PII without authorization." });

  return out;
}

export function summarize(checks: ComplianceCheck[], atp?: number, uaopVersion = "1.0.0"): ComplianceStatus {
  const violations = checks.filter((c) => !c.passed);
  const status: ComplianceStatus["status"] =
    violations.length === 0 ? "certified" : violations.length <= 2 ? "degraded" : "suspended";
  return {
    status,
    atp,
    uaop_version: uaopVersion,
    active_violations: violations.map((v) => `UAOP Article ${v.article}: ${v.rule}`),
    last_updated: new Date().toISOString(),
  };
}

File: sdk/python/meridian/compliance.py
"""Compliance Monitor — per UAOP article."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import List, Optional


@dataclass
class CallContext:
    principal: Optional[dict] = None
    declared_capabilities: Optional[List[str]] = None
    requested_action: Optional[str] = None
    is_irreversible: Optional[bool] = None
    has_auth_token: Optional[bool] = None
    human_asked_identity: Optional[bool] = None
    agent_disclosed: Optional[bool] = None
    used_tokens: Optional[int] = None
    budget_tokens: Optional[int] = None
    pii: Optional[bool] = None
    pii_authorized: Optional[bool] = None
    claim_confidence: Optional[float] = None
    domain: Optional[str] = None


@dataclass
class ComplianceCheck:
    article: int
    passed: bool
    rule: str
    note: Optional[str] = None


def check(ctx: CallContext) -> List[ComplianceCheck]:
    out: List[ComplianceCheck] = []

    out.append(
        ComplianceCheck(
            article=1,
            rule="Claims must carry a confidence score >= 0.3.",
            passed=ctx.claim_confidence is None or ctx.claim_confidence >= 0.3,
        )
    )

    in_scope = (
        not ctx.requested_action
        or not ctx.declared_capabilities
        or any(c in ctx.requested_action for c in ctx.declared_capabilities)
    )
    out.append(
        ComplianceCheck(
            article=2,
            rule="Requested action must match a declared capability.",
            passed=in_scope,
        )
    )

    out.append(ComplianceCheck(article=3, rule="No known conflict among principals (SDK-level check).", passed=True))

    irreversible_ok = not ctx.is_irreversible or ctx.has_auth_token is True
    out.append(
        ComplianceCheck(
            article=4,
            rule="Irreversible actions require an explicit authorization token.",
            passed=irreversible_ok,
        )
    )

    within_budget = not ctx.budget_tokens or (ctx.used_tokens or 0) <= ctx.budget_tokens
    out.append(
        ComplianceCheck(article=5, rule="Stay within per-task token budget.", passed=within_budget)
    )

    disclosed = not ctx.human_asked_identity or ctx.agent_disclosed is True
    out.append(
        ComplianceCheck(article=6, rule="Disclose AI nature when directly asked.", passed=disclosed)
    )

    pii_ok = not ctx.pii or ctx.pii_authorized is True
    out.append(
        ComplianceCheck(article=7, rule="Do not transmit PII without authorization.", passed=pii_ok)
    )

    return out


def summarize(checks: List[ComplianceCheck], atp: Optional[int] = None, uaop_version: str = "1.0.0") -> dict:
    violations = [c for c in checks if not c.passed]
    status = "certified" if not violations else ("degraded" if len(violations) <= 2 else "suspended")
    return {
        "status": status,
        "atp": atp,
        "uaop_version": uaop_version,
        "active_violations": [f"UAOP Article {v.article}: {v.rule}" for v in violations],
        "last_updated": datetime.now(timezone.utc).isoformat(),
    }
